### ISSUE TITLE
Restore orginal domain object from persisted snapshot

### ISSUE THREAD

**Issue by eladh** [javers/javers#133]
Currently there is no possiblity to retrive original object or converte it from CDO.
In order to produce some kind of POJO based version management Eco-system we must be able to retrieve the original POJO.

Generally speaking i would love to see Javers  with version management API like :
1. Retrieve the Original POJO
2. Merge Operation => with Updated/Conflict status 
3. Restore Object to X version 

do you Store all Object graph or just Incremental information / diff from the last commit ?

10x

**Comment by akarmazin** [comment]
1) For retrieving the original POJO as a quick and dirty solution you could use the class below (or something similar). 
see loadLatestEntityState() or loadEntityStateFromSnapshot() or loadEntityStateForCommitId()
However, as for now this solution has some restrictions (see class level comments).
2) For analyzing the conflict status you has to remember the commitId (and/or @Version value if exists) on POJO load operation in order to retrieve your base entity. By commiting the changed entity to your database you have to analyze the changes you maid against the base entity and changes already done by another user (actual entity state): here you could use Javers.compare for deep comparison, however the solution in case of deeply located elements in object graph would be not trivial
3) Restoring the whole object graph in relational database is not a trivial operation as well and mostly (or ever) is highly dependent on your business logic ->  So I don't think Javers could provide any general solution for this operation.

```
package javers.repository.test.reader;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.javers.common.collections.Optional;
import org.javers.common.reflection.JaversField;
import org.javers.core.Javers;
import org.javers.core.commit.CommitId;
import org.javers.core.metamodel.object.CdoSnapshot;
import org.javers.core.metamodel.object.InstanceId;
import org.javers.core.metamodel.object.InstanceIdDTO;
import org.javers.core.metamodel.property.Property;

/**
 * Loads under certain assumptions the original entity state for given
 * CdoSnapshot or CommitId.<br>
 * 1) Loads the whole object graph eagerly (lazy loading would require the use
 * of proxy objects). <br>
 * 2) All entities have constructor with no arguments -> Java Beans standard. <br>
 * 3) Each entity field ({@link JaversField}) has a setter method with the
 * argument of the same type as field -> Java Beans standard. <br>
 * 4) Since Javers lacks the function for retrieving the CdoSnapshot based on
 * nearest commitId -> all CdoSnapshots have to be loaded (assumes that limit =
 * -1 means all snapshots, see implementation of
 * {@linkplain #loadEntityStateForCommitId(InstanceId, String)}). <br>
 * 5) Maps and other types are not yet supported. <br>
 */
public class JaversRepositoryReader {

	private static final double EPS = 0.001;
	private Javers javers;

	private Map<Object, Object> idToEntityMap = new HashMap<>();

	public JaversRepositoryReader(Javers javers) {
		this.javers = javers;
	}

	public void clearCache() {
		idToEntityMap.clear();
	}

	/**
	 * Loads original entity state for given CdoSnapshot
	 * 
	 * @param snapshot
	 * @return
	 */
	public Object loadEntityStateFromSnapshot(CdoSnapshot snapshot) {
		clearCache();
		try {
			return loadEntityInternal(snapshot);
		} catch (IllegalAccessException | InstantiationException e) {
			e.printStackTrace();
			return null;
		}
	}

	/**
	 * Loads original entity state for given instanceId and commitId
	 * 
	 * @param instanceId
	 * @param commitId
	 * @return
	 */
	public Object loadEntityStateForCommitId(InstanceId instanceId,
			String commitId) {
		clearCache();
		try {
			return loadEntityForCommitId(instanceId, CommitId.valueOf(commitId));
		} catch (IllegalAccessException | InstantiationException e) {
			e.printStackTrace();
			return null;
		}
	}

	/**
	 * Loads the latest commited entity state for given instanceId
	 * 
	 * @param instanceId
	 * @return
	 */
	public Object loadLatestEntityState(InstanceId instanceId) {
		Optional<CdoSnapshot> latestSnapshot = javers
				.getLatestSnapshot(convertInstanceIdToDTO(instanceId));
		return latestSnapshot.isPresent() ? loadEntityStateFromSnapshot(latestSnapshot
				.get()) : null;
	}

	/**
	 * Load latest entity state for entity
	 * 
	 * @param entity
	 * @return
	 */
	public Object loadLatestEntityStateForEntity(Object entity) {
		return loadLatestEntityState(javers.idBuilder().instanceId(entity));
	}

	@SuppressWarnings({ "rawtypes", "unchecked" })
	private Object loadEntityInternal(CdoSnapshot snap)
			throws IllegalAccessException, InstantiationException {

		if (idToEntityMap.containsKey(snap.getGlobalId().getCdoId()))
			return idToEntityMap.get(snap.getGlobalId().getCdoId());

		Class clientsClass = snap.getManagedClass().getClientsClass();
		Object instance = clientsClass.newInstance();
		idToEntityMap.put(snap.getGlobalId().getCdoId(), instance);
		for (Property property : snap.getProperties()) {
			Object propertyValue = snap.getPropertyValue(property);
			Object propertyValueToSet = propertyValue;
			if (propertyValue != null) {
				try {
					Method setter = getSetterForProperty(clientsClass, property);
					Class<?> targetType = setter.getParameterTypes()[0];
					if (targetType.isAssignableFrom(java.util.List.class)
							|| targetType.isAssignableFrom(java.util.Set.class)) {
						Collection collection = (Collection) propertyValue;
						Set toBeDeleted = new HashSet<>();
						for (Object element : collection) {
							if (element instanceof InstanceId) {
								InstanceId instanceId = (InstanceId) element;
								Object targetObject = loadEntityForCommitId(
										instanceId, snap.getCommitId());
								collection.add(targetObject);
								toBeDeleted.add(element);
							}
						}
						for (Object element : toBeDeleted) {
							collection.remove(element);
						}
					} else if (propertyValue instanceof InstanceId) {
						propertyValueToSet = loadEntityForCommitId(
								(InstanceId) propertyValue, snap.getCommitId());
					}
					setter.invoke(instance, propertyValueToSet);
				} catch (Exception exc) {
					exc.printStackTrace();
				}
			}
		}

		return instance;
	}

	/**
	 * Programmatic way is chosen to find corresponding CdoSnapshot, however
	 * better practice would be to have the repository method for searching
	 * 
	 * @param instanceId
	 * @param commitId
	 * @return
	 * @throws InstantiationException
	 * @throws IllegalAccessException
	 */
	private Object loadEntityForCommitId(InstanceId instanceId,
			CommitId commitId) throws IllegalAccessException,
			InstantiationException {

		// TODO here the Javers function for loading the CdoSnapshot for the
		// given commitId should be called, but the current Javers
		// implementation doesn't provide it
		// TODO assumes that limit = -1 means "all CdoSnapshots"
		List<CdoSnapshot> snapshots = javers.getStateHistory(
				convertInstanceIdToDTO(instanceId), -1); // load all

		double commitToBeSearched = commitIdAsDouble(commitId);
		CdoSnapshot snapFound = null;
		// since all snapshot are sorted from the highest commit value to the
		// lowest one
		for (CdoSnapshot snap : snapshots) {
			if (commitIdAsDouble(snap.getCommitId()) < (commitToBeSearched + EPS)) {
				snapFound = snap;
			}
		}

		Object instance = loadEntityInternal(snapFound);

		return instance;
	}

	private InstanceIdDTO convertInstanceIdToDTO(InstanceId instanceId) {
		return InstanceIdDTO.instanceId(instanceId.getCdoId(), instanceId
				.getCdoClass().getClientsClass());
	}

	private double commitIdAsDouble(CommitId commitId) {
		return Double.parseDouble(commitId.value());
	}

	@SuppressWarnings("rawtypes")
	private Method getSetterForProperty(Class clientsClass, Property property)
			throws NoSuchFieldException, NoSuchMethodException {
		Field declaredField = clientsClass.getDeclaredField(property.getName());
		@SuppressWarnings("unchecked")
		Method setter = clientsClass.getDeclaredMethod(
				getSetterNameForProperty(property), declaredField.getType());
		return setter;
	}

	private String getSetterNameForProperty(Property property) {
		return "set" + property.getName().substring(0, 1).toUpperCase()
				+ property.getName().substring(1);
	}

}
```

**Comment by bartoszwalacik** [comment]
well, the bigest problem when we try to restore objects from JaversRepository are references.
Restoring a simple ValueObject is relatively easey, but usually we have a object graph. When you restore one node, you may want to follow references to other objects.
This requires tricky lazy-loading proxies like those in Hibernate ... lot of work

**Comment by eladh** [comment]
thanks for the explanation :-) 
is there any chance for a roadmap  - we really want to integrate Javers in our product but needs this features as core functionality.

10x

**Comment by bartoszwalacik** [comment]
hi, now we are finishing implementing JQL (Javers Query Language)
which is a really big step forward.

Concerning this issue, could you give more details about your requirements?
As I said before, restoring a simple, isolated ValueObject (object without references) is quite easy.
But when you want to restore a whole object graph, things gets complicated ...

**Comment by eladh** [comment]
Hi :-)
i will check JaversRepositoryReader.
for now i will commit the changes by @Entity  and isolate the whole object graph .
will try to read the project source code and hopefully write a design doc about it.

10x
elad

**Comment by bartoszwalacik** [comment]
ok, so we stay in touch, in case if you would like to join our new Javers gitter chat room, we are here:
https://gitter.im/javers/javers

**Comment by bgalek** [comment]
first step: we will provide a method to get graph with limited depth, eagerly loaded
ToDo:
- create InstanceCreator interface available for beeing provided by user
- javers should try to make object by itself if class has empty constructor

**Comment by smirnp** [comment]
The code above is great, but a little bit obsolete and buggy (the suppressed "Сurrent modiаication exception" of collection in (Object element : collection){...} cycle.

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import States.ClusterState;
import org.javers.common.collections.Optional;
import org.javers.common.reflection.JaversField;
import org.javers.core.Javers;
import org.javers.core.commit.CommitId;
import org.javers.core.metamodel.object.CdoSnapshot;
import org.javers.core.metamodel.object.InstanceId;
//import org.javers.core.metamodel.object.InstanceIdDTO;
import org.javers.repository.jql.InstanceIdDTO;
import org.javers.core.metamodel.property.Property;
import org.javers.repository.jql.QueryBuilder;

/**
 * Loads under certain assumptions the original entity state for given
 * CdoSnapshot or CommitId.<br>
 * 1) Loads the whole object graph eagerly (lazy loading would require the use
 * of proxy objects). <br>
 * 2) All entities have constructor with no arguments -> Java Beans standard. <br>
 * 3) Each entity field ({@link JaversField}) has a setter method with the
 * argument of the same type as field -> Java Beans standard. <br>
 * 4) Since Javers lacks the function for retrieving the CdoSnapshot based on
 * nearest commitId -> all CdoSnapshots have to be loaded (assumes that limit =
 * -1 means all snapshots, see implementation of
 * {@linkplain #loadEntityStateForCommitId(InstanceId, String)}). <br>
 * 5) Maps and other types are not yet supported. <br>
 */
public class JaversRepositoryReader {

    private static final double EPS = 0.001;
    private Javers javers;

    private Map<Object, Object> idToEntityMap = new HashMap<>();

    public JaversRepositoryReader(Javers javers) {
        this.javers = javers;
    }

    public void clearCache() {
        idToEntityMap.clear();
    }

    /**
     * Loads original entity state for given CdoSnapshot
     *
     * @param snapshot
     * @return
     */
    public Object loadEntityStateFromSnapshot(CdoSnapshot snapshot) {
        clearCache();
        try {
            return loadEntityInternal(snapshot);
        } catch (IllegalAccessException | InstantiationException e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * Loads original entity state for given instanceId and commitId
     *
     * @param instanceId
     * @param commitId
     * @return
     */
    public Object loadEntityStateForCommitId(InstanceId instanceId,
                                             String commitId) {
        clearCache();
        try {
            return loadEntityForCommitId(instanceId, CommitId.valueOf(commitId));
        } catch (IllegalAccessException | InstantiationException e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * Loads the latest commited entity state for given instanceId
     *
     * @param instanceId
     * @return
     */
    public Object loadLatestEntityState(InstanceId instanceId) {
        Optional<CdoSnapshot> latestSnapshot = javers.getLatestSnapshot(convertInstanceIdToDTO(instanceId));
        //Optional<CdoSnapshot> latestSnapshot = javers.getLatestSnapshot()getLatestSnapshot(convertInstanceIdToDTO(instanceId));
        return latestSnapshot.isPresent() ? loadEntityStateFromSnapshot(latestSnapshot
                .get()) : null;
    }

    /**
     * Load latest entity state for entity
     *
     * @param entity
     * @return
     */
    public Object loadLatestEntityStateForEntity(Object entity) {
        return loadLatestEntityState(javers.idBuilder().instanceId(entity));
    }

    @SuppressWarnings({ "rawtypes", "unchecked" })
    private Object loadEntityInternal(CdoSnapshot snap)
            throws IllegalAccessException, InstantiationException {

        if (idToEntityMap.containsKey(snap.getGlobalId().getCdoId()))
            return idToEntityMap.get(snap.getGlobalId().getCdoId());

        Class clientsClass = snap.getManagedType().getBaseJavaClass();

        Object instance = clientsClass.newInstance();
        idToEntityMap.put(snap.getGlobalId().getCdoId(), instance);
        for (Property property : snap.getProperties()) {
            Object propertyValue = snap.getPropertyValue(property);
            Object propertyValueToSet = propertyValue;
            if (propertyValue != null) {
                try {
                    Method setter = getSetterForProperty(clientsClass, property);
                    Class<?> targetType = setter.getParameterTypes()[0];
                    if (targetType.isAssignableFrom(java.util.List.class)
                            || targetType.isAssignableFrom(java.util.Set.class)) {
                        Collection collection = (Collection) propertyValue;
                        Set toBeDeleted = new HashSet<>();
                        for (Object element : collection.toArray()){
                            if (element instanceof InstanceId) {
                                InstanceId instanceId = (InstanceId) element;

                                Object targetObject = loadEntityForCommitId(
                                        instanceId, snap.getCommitId());
                                collection.add(targetObject);
                                toBeDeleted.add(element);
                             }
                        }
                        for (Object element : toBeDeleted) {
                            collection.remove(element);
                        }
                    } else if (propertyValue instanceof InstanceId) {
                        propertyValueToSet = loadEntityForCommitId(
                                (InstanceId) propertyValue, snap.getCommitId());
                    }
                    setter.invoke(instance, propertyValueToSet);
                } catch (Exception exc) {
                    exc.printStackTrace();
                }
            }
        }

        return instance;
    }

    /**
     * Programmatic way is chosen to find corresponding CdoSnapshot, however
     * better practice would be to have the repository method for searching
     *
     * @param instanceId
     * @param commitId
     * @return
     * @throws InstantiationException
     * @throws IllegalAccessException
     */
    private Object loadEntityForCommitId(InstanceId instanceId,
                                         CommitId commitId) throws IllegalAccessException,
            InstantiationException {

        // TODO here the Javers function for loading the CdoSnapshot for the
        // given commitId should be called, but the current Javers
        // implementation doesn't provide it
        // TODO assumes that limit = -1 means "all CdoSnapshots"
        //List<CdoSnapshot> snapshots = javers.getStateHistory(convertInstanceIdToDTO(instanceId), -1); // load all
        List<CdoSnapshot> snapshots = javers.findSnapshots(QueryBuilder.byInstanceId(instanceId, instanceId.getManagedType().getBaseJavaClass() ).build());

        double commitToBeSearched = commitIdAsDouble(commitId);
        CdoSnapshot snapFound = null;
        // since all snapshot are sorted from the highest commit value to the
        // lowest one
        for (CdoSnapshot snap : snapshots) {
            if (commitIdAsDouble(snap.getCommitId()) < (commitToBeSearched + EPS)) {
                snapFound = snap;
            }
        }

        Object instance = loadEntityInternal(snapFound);

        return instance;
    }

    private InstanceIdDTO convertInstanceIdToDTO(InstanceId instanceId) {
        return InstanceIdDTO.instanceId(instanceId.getCdoId(), instanceId.getManagedType().getBaseJavaClass());//..getClientsClass());
    }

    private double commitIdAsDouble(CommitId commitId) {
        return Double.parseDouble(commitId.value());
    }

    @SuppressWarnings("rawtypes")
    private Method getSetterForProperty(Class clientsClass, Property property)
            throws NoSuchFieldException, NoSuchMethodException {
        Field declaredField = clientsClass.getDeclaredField(property.getName());
        @SuppressWarnings("unchecked")
        Method setter = clientsClass.getDeclaredMethod(
                getSetterNameForProperty(property), declaredField.getType());
        return setter;
    }

    private String getSetterNameForProperty(Property property) {
        return "set" + property.getName().substring(0, 1).toUpperCase()
                + property.getName().substring(1);
    }

}

**Comment by bartoszwalacik** [comment]
@smirnp, it would be better to have this code distributed with JaVers. We can accept Pull Request with this feature, with incubating status, to separate JaVers module, lets say, javers-instance-creator. Would you like to contribute this way?

**Comment by smirnp** [comment]
Ok, I'll do it

**Comment by smirnp** [comment]
I'll do it

**Comment by enietom** [comment]
Hi there, I took the class sample you added previously and made some modification to made it work correctly (at least for my project).
Hope it helps:

`public class JaversSnapshotsCompiler {

	private Log log = LogFactory.getLog(getClass());

	private Javers javers;

	private Map<Object, Object> idToEntityMap = new HashMap<>();

	public JaversSnapshotsCompiler(Javers javers) {
		this.javers = javers;
	}

	public void clearCache() {
		idToEntityMap.clear();
	}

	/**
	 * Loads original entity state for given CdoSnapshot
	 * 
	 * @param snapshot
	 * @return
	 */
	public Object compileEntityStateFromSnapshot(CdoSnapshot snapshot) {
		clearCache();
		try {
			return compileEntityInternal(snapshot, null);
		} catch (IllegalAccessException | InstantiationException e) {
			log.error("Error in compileEntityInternal. For snapshot: " + snapshot, e);
			return null;
		}
	}

	/**
	 * Loads original entity state for given instanceId and commitId
	 * 
	 * @param instanceId
	 * @param commitId
	 * @return
	 */
	public Object compileEntityStateForCommitId(InstanceId instanceId, String commitId) {
		clearCache();
		try {
			return compileEntityForCommitId(instanceId, CommitId.valueOf(commitId));
		} catch (IllegalAccessException | ClassNotFoundException | InstantiationException e) {
			log.error("Error in compileEntityForCommitId. For instanceId: " + instanceId, e);
			return null;
		}
	}

	/**
	 * Loads the latest commited entity state for given instanceId
	 * 
	 * @param instanceId
	 * @return
	 */
	public Object compileLatestEntityState(InstanceId instanceId) {
		InstanceIdDTO dto = convertInstanceIdToDTO(instanceId);
		Optional<CdoSnapshot> latestSnapshot = javers.getLatestSnapshot(dto, dto.getClass());
		return latestSnapshot.isPresent() ? compileEntityStateFromSnapshot(latestSnapshot.get()) : null;
	}

	/**
	 * Load latest entity state for entity
	 * 
	 * @param entity
	 * @return
	 */
	public Object compileLatestEntityStateForEntity(Object entity) {
		InstanceId instanceId = javers.idBuilder().instanceId(entity);
		return compileLatestEntityState(instanceId);
	}

	@SuppressWarnings({ "rawtypes", "unchecked" })
	private Object compileEntityInternal(CdoSnapshot snap, CommitId currentCommitId) throws IllegalAccessException, InstantiationException {
		if (snap.getGlobalId() instanceof InstanceId
				&& idToEntityMap.containsKey(((InstanceId) snap.getGlobalId()).value())) {
			return idToEntityMap.get(((InstanceId) snap.getGlobalId()).value());
		}

		Class clientsClass = snap.getManagedType().getBaseJavaClass();
		Object instance = clientsClass.newInstance();
		idToEntityMap.put(((InstanceId) snap.getGlobalId()).value(), instance);
		
		if (currentCommitId == null) {
			currentCommitId = snap.getCommitId();
		}

		for (String propertyName : snap.getState().getProperties()) {
			Object propertyValue = snap.getPropertyValue(propertyName);
			Object propertyValueToSet = propertyValue;
			if (propertyValue != null) {
				try {
					Method setter = getSetterForProperty(clientsClass, propertyName);

					try {
						/*
						 * Try to process property as collection.
						 * if(targetType.isAssignableFrom(List.class) ||
						 * targetType.isAssignableFrom(Set.class)) not always
						 * defines correctly
						 */
						Collection collection = (Collection) propertyValue;
						Set toBeDeleted = new HashSet<>();
						for (Object element : collection.toArray()) {
							if (element instanceof InstanceId) {
								InstanceId instanceId = (InstanceId) element;

								Object targetObject = compileEntityForCommitId(instanceId, currentCommitId);
								collection.add(targetObject);
								toBeDeleted.add(element);
							}
						}
						for (Object element : toBeDeleted) {
							collection.remove(element);
						}

						propertyValueToSet = collection;
					} catch (Exception exc) {
						if (propertyValue instanceof InstanceId) {
							propertyValueToSet = compileEntityForCommitId((InstanceId) propertyValue,
									currentCommitId);
						} else {
							// some primitive type
							propertyValueToSet = propertyValue;
						}
					}

					setter.invoke(instance, propertyValueToSet);
				} catch (Exception exc) {
					log.error("Error using reflections API in getSetterForProperty. For snap: " + snap, exc);
				}
			}
		}

		return instance;
	}

	/**
	 * Programmatic way is chosen to find corresponding CdoSnapshot, however
	 * better practice would be to have the repository method for searching
	 * 
	 * @param instanceId
	 * @param commitId
	 * @return
	 * @throws InstantiationException
	 * @throws IllegalAccessException
	 * @throws ClassNotFoundException
	 */
	private Object compileEntityForCommitId(InstanceId instanceId, CommitId commitId) throws IllegalAccessException,
			InstantiationException, ClassNotFoundException {

		List<CdoSnapshot> snapshots = javers.findSnapshots(QueryBuilder.byInstanceId(instanceId.getCdoId(),
				Class.forName(instanceId.getTypeName())).build());

		CdoSnapshot lastSnap = null;
		CdoSnapshot commitSnap = null;

		/*
		 * Snapshots are ordered from the newest to the eldest. So the last
		 * version of a snap is the first found on the snapshots list
		 */
		for (CdoSnapshot snap : snapshots) {
			if (lastSnap == null) {
				lastSnap = snap;
			}
			if (snap.getCommitId().equals(commitId)) {
				commitSnap = snap;
				break;
			}
		}

		Object instance = compileEntityInternal((commitSnap != null) ? commitSnap : lastSnap, commitId);

		return instance;
	}

	private InstanceIdDTO convertInstanceIdToDTO(InstanceId instanceId) {
		return InstanceIdDTO.instanceId(instanceId.getCdoId(), instanceId.getClass());
	}

	@SuppressWarnings({ "rawtypes", "unchecked" })
	private Method getSetterForProperty(Class clientsClass, String propertyName) throws NoSuchFieldException,
			NoSuchMethodException {
		Method getter = null;
		try {
			getter = clientsClass.getMethod(getGetterNameForProperty(propertyName, false));
		} catch (NoSuchMethodException e) {
			/*
			 * For boolean, if I could find a "getProperty" try "isProperty"
			 * methods
			 */
			getter = clientsClass.getMethod(getGetterNameForProperty(propertyName, true));
		}
		Method setter = clientsClass.getMethod(getSetterNameForProperty(propertyName), getter.getReturnType());
		return setter;
	}

	private String getSetterNameForProperty(String propertyName) {
		return "set" + propertyName.substring(0, 1).toUpperCase() + propertyName.substring(1);
	}

	private String getGetterNameForProperty(String propertyName, boolean isBoolean) {
		String prefix = (isBoolean) ? "is" : "get";
		return prefix + propertyName.substring(0, 1).toUpperCase() + propertyName.substring(1);
	}

}`

**Comment by bartoszwalacik** [comment]
We have decided to design SnapshotsCompiler feature in a different way. Snapshots will be converted to JSON and then we will use GSon deserialization

**Comment by milanov** [comment]
Hi @bartoszwalacik , do you have an approximate time frame when this will be implemented, since we've done an incomplete bug-ridden version of this "entity restore" feature and could use a built-in version

**Comment by bartoszwalacik** [comment]
Well, hope it will be released in summer

**Comment by bartoszwalacik** [comment]
this looks like a quite big feature in JaVers. Some support from community would be helpful ... is anyone here eager to contribute?

**Comment by rasheedamir** [comment]
Hi @bartoszwalacik pretty nice framework! Is there any update on this?

**Comment by bartoszwalacik** [comment]
No one is working on it.

**Comment by bartoszwalacik** [comment]
@rasheedamir
I'm thinking about developing this feature. Maybe it will be the next big thing in JaVers ...

**Comment by rasheedamir** [comment]
awesome @bartoszwalacik let me know if need assistance; I mean if you have a plan how to implement then I can offer assistance either myself or through one of my team members

**Comment by bartoszwalacik** [comment]
Good to hear @rasheedamir, your contributions will help us a lot.
I've created a dedicated chat room for this project https://gitter.im/javers/shadows
Please join this chat for quick collaboration.

So how we start?
First, I've already pushed some code to this branch https://github.com/javers/javers/tree/shadows 
The simplest case is implemented, take a look at the new test - ShadowFactoryTest.groovy 

## Design
Looking from users perspective, we are going to add one method to javers:

```
List<Object> findShadows(JqlQuery query)
```
**Shadow** means users domain object restored from a Snapshot.
We need to add a switch to the query

* shallow shadows - no references reslolved, identifiers to other entities nulled, this query will be fast and implementation will be simple (run findSnapshots() and convert all Snapshots to Shadows, ignoring references)
* deep shadows - shadows will be created deeply, following references (like Hibernate lazy loading). Reference resolving scope will be limited to original commit scope. This query will be mush slower

### ShadowFactory implementation

**Comment by bartoszwalacik** [comment]
done and released in 3.2.0
see http://javers.org/documentation/jql-examples/#query-for-shadows
